clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -60,67 +68,165 @@
+        }
+
+        public short code() {
+            return code;
+        }
+
+        public static Errors fromCode(final short code) {
+            switch (code) {
+                case 0:
+                    return NONE;
+                case 1:
+                    return NEED_REJOIN;
+                default:
+                    throw new IllegalArgumentException("Unknown error code: " + code);
+            }
+        }
+    }
+
+    public static ByteBuffer serializeSubscriptionV0(PartitionAssignor.Subscription subscription) {
         Struct struct = new Struct(SUBSCRIPTION_V0);
         struct.set(USER_DATA_KEY_NAME, subscription.userData());
         struct.set(TOPICS_KEY_NAME, subscription.topics().toArray());
+
         ByteBuffer buffer = ByteBuffer.allocate(CONSUMER_PROTOCOL_HEADER_V0.sizeOf() + SUBSCRIPTION_V0.sizeOf(struct));
         CONSUMER_PROTOCOL_HEADER_V0.writeTo(buffer);
         SUBSCRIPTION_V0.write(buffer, struct);
         buffer.flip();
         return buffer;
     }
 
-    public static PartitionAssignor.Subscription deserializeSubscription(ByteBuffer buffer) {
-        Struct header = CONSUMER_PROTOCOL_HEADER_SCHEMA.read(buffer);
-        Short version = header.getShort(VERSION_KEY_NAME);
-        checkVersionCompatibility(version);
+    public static ByteBuffer serializeSubscriptionV1(PartitionAssignor.Subscription subscription) {
+        Struct struct = new Struct(SUBSCRIPTION_V1);
+        struct.set(USER_DATA_KEY_NAME, subscription.userData());
+        struct.set(TOPICS_KEY_NAME, subscription.topics().toArray());
+        List<Struct> topicAssignments = new ArrayList<>();
+        Map<String, List<Integer>> partitionsByTopic = CollectionUtils.groupPartitionsByTopic(subscription.ownedPartitions());
+        for (Map.Entry<String, List<Integer>> topicEntry : partitionsByTopic.entrySet()) {
+            Struct topicAssignment = new Struct(TOPIC_ASSIGNMENT_V0);
+            topicAssignment.set(TOPIC_KEY_NAME, topicEntry.getKey());
+            topicAssignment.set(PARTITIONS_KEY_NAME, topicEntry.getValue().toArray());
+            topicAssignments.add(topicAssignment);
+        }
+        struct.set(OWNED_PARTITIONS_KEY_NAME, topicAssignments.toArray());
+
+        ByteBuffer buffer = ByteBuffer.allocate(CONSUMER_PROTOCOL_HEADER_V1.sizeOf() + SUBSCRIPTION_V1.sizeOf(struct));
+        CONSUMER_PROTOCOL_HEADER_V1.writeTo(buffer);
+        SUBSCRIPTION_V1.write(buffer, struct);
+        buffer.flip();
+        return buffer;
+    }
+
+    public static ByteBuffer serializeSubscription(PartitionAssignor.Subscription subscription) {
+        switch (subscription.version()) {
+            case CONSUMER_PROTOCOL_V0:
+                return serializeSubscriptionV0(subscription);
+
+            case CONSUMER_PROTOCOL_V1:
+                return serializeSubscriptionV1(subscription);
+
+            default:
+                // for any versions higher than known, try to serialize it as V1
```

Comment:
More out of curiosity, but I'm assuming we'll always try to serialize to the latest version, but why not throw an exception? Seems it would be a bug if we hit this point.  Same applies to deserializing below.

Reply:
We want to leave the door open for "future proof": e.g. in the current code https://github.com/apache/kafka/pull/6528/files/e54d8e197bd593369b51f2f2e07a7f120c8cf606#diff-bad29ccb1aba700e1badeff62f1a86b7L147

We just check if the version is not negative, and if not we always assume it can be deserialized as v0; this is indeed true still: say old byte-code gets a v1 subscription object, it can still successfully deserialize as v0.